### PR TITLE
Add OutputType property to the BrowseObject so that the application prop page can open

### PR DESCRIPTION
**Pull request by srivatsn**
The application property page was getting the OutputType of the application through an interface that vslangproj implements and CPS doesn't. Instead, we will get the property from the BrowseObject. I've duplicated the XAML definitions for the BrowseObject and made it override the one in the MSBuild directory so that we can add to it. What we do with the MSBuild one is yet to be figured out. 

OutputType property in MSBuild maps to OutputTypeEx in vslangproj and needs to be integers. OutputType itself is a smaller list and CPS doesn't like Enums with repeated values in the XAML file. So I've implemented a IDynamicEnumProvider to do the mapping from OutputTypeEx to OutputType.

Tagging @dotnet/roslyn-analysis

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/None.xaml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  <StringProperty Name="Link" Visible="false" />
+  <StringProperty Name="Extension" Visible="False" ReadOnly="true">
+    <StringProperty.DataSource>
+      <DataSource Persistence="Intrinsic" ItemType="None" PersistedName="Extension" />
+    </StringProperty.DataSource>
+  </StringProperty>
+  <StringProperty Name="LastGenOutput" Visible="false" Description="The filename of the last file generated as a result of the SFG." />
+  <BoolProperty Name="DesignTime" Visible="false" Description="A value indicating whether this file has a designer." />
+  <BoolProperty Name="AutoGen" Visible="false" Description="A value indicating whether this is a generated file." />
+  <StringProperty Name="CustomTool" Visible="false" Description="DTE Property for accessing the Generator property.">
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/None.xaml]: This is the primary change to this file needed for the Settings designer to work. Including it in this PR.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      <DataSource PersistedName="RootNamespace" Persistence="ProjectFile" Label="Configuration" HasConfigurationCondition="false" />
+    </StringProperty.DataSource>
+  </StringProperty>
+  <StringProperty Name="TargetFramework" Visible="False">
+    <StringProperty.DataSource>
+      <DataSource Persistence="ProjectFile" PersistedName="TargetFrameworkMoniker" />
+    </StringProperty.DataSource>
+  </StringProperty>
+  <StringProperty Name="OutputName" Visible="False" />
+  <DynamicEnumProperty Name="OutputType" DisplayName="Output Type" EnumProvider="OutputTypeEnumProvider" />
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: From here, until Win32Resource file below are the new properties that I've added to this file - everything else is that same as the one in MSBuild.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--Copyright, Microsoft Corporation, All rights reserved.-->
+<Rule
+  Name="ConfigurationGeneralBrowseObject"
+  DisplayName="General"
+  PageTemplate="generic"
+  Description="General"
+  OverrideMode= "Replace"
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: This lets us override the definition in MSBuild.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class OutputTypeEnumProvider : IDynamicEnumValuesProvider
+    {
+        public Task<IDynamicEnumValuesGenerator> GetProviderAsync(IList<NameValuePair> options)
+        {
+            return Task.FromResult<IDynamicEnumValuesGenerator>(new OutputTypeEnumValuesGenerator());
+        }
+
+        internal class OutputTypeEnumValuesGenerator : IDynamicEnumValuesGenerator
+        {
```
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: What does DisplayName convey here ? Name and Display name doesn't look consistent

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      <DataSource Persistence="ProjectFile" PersistedName="TargetFrameworkMoniker" />
+    </StringProperty.DataSource>
+  </StringProperty>
+  <StringProperty Name="OutputName" Visible="False" />
+  <DynamicEnumProperty Name="OutputType" DisplayName="Output Type" EnumProvider="OutputTypeEnumProvider" />
+  <EnumProperty Name="OutputTypeEx" DisplayName="Output Type">
+    <EnumValue Name="dll" DisplayName="0" />
+    <EnumValue Name="exe" DisplayName="1" />
+    <EnumValue Name="winexe" DisplayName="2" />
+    <EnumValue Name="WinMDObj" DisplayName="3" />
```
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Should WinMDObj be lower case here , or it doesn't matter since OutPutTypeEnumProvider defines the right set ?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    <EnumValue Name="appcontainerexe" DisplayName="4" />
+    <EnumProperty.DataSource>
+      <DataSource Persistence="ProjectFile" PersistedName="OutputType" />
+    </EnumProperty.DataSource>
+  </EnumProperty>
+  <StringProperty Name="StartupObject" DisplayName="Type that contains the entry point" />
```
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Did you add the StartupObject ? why is that needed ?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 5 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: I'll lower case it for consistency but it doesn't seem to matter. Looks like CPS does a case-insensitive comparison. No for OutputTypeEx the enum provider doesn't play any role. It just comes from this enum.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 6 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Not needed for this PR. I was just enabling more stuff in the application property page. (This and Win32ResourceFile below is new and are needed by the application property page).

File: src/Targets/Microsoft.CSharp.DesignTime.targets [hunk 7]
```diff
@@ -26,6 +26,13 @@
     <PropertyPageSchema Include="$(CustomCommonXamlResourcesDirectory)AppDesigner.xaml">
       <Context>ProjectSubscriptionService</Context>
     </PropertyPageSchema>
+  
+    <PropertyPageSchema Include="$(CustomCommonXamlResourcesDirectory)General.BrowseObject.xaml">
+      <Context>BrowseObject</Context>
+    </PropertyPageSchema>
+  
+    <PropertyPageSchema Include="$(CustomCommonXamlResourcesDirectory)None.xaml">
+      <Context>File;BrowseObject</Context>
```
Comment (jinujoseph) [inline src/Targets/Microsoft.CSharp.DesignTime.targets]: Any reason why context for this is file too for none.xaml

**Comment by jinujoseph** [comment]
LGTM

File: src/Targets/Microsoft.CSharp.DesignTime.targets (hunk 7 above)
Comment (srivatsn) [inline src/Targets/Microsoft.CSharp.DesignTime.targets]: The Context tells CPS the services to which these properties should be published. All the item related ones are published to File as well. I don't know how to get at them yet.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj [hunk 8]
```diff
@@ -95,6 +96,14 @@
       <Generator>MSBuild:GenerateRuleSourceFromXaml</Generator>
       <SubType>Designer</SubType>
     </XamlPropertyRule>
+    <XamlPropertyRule Include="ProjectSystem\Rules\General.BrowseObject.xaml">
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj]: Can you follow the same pattern for the other rules where we add a partial class and mark it with "ExcludeFromCodeCoverage"? We don't cover properties.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            private readonly Dictionary<string, IEnumValue> _outputTypeValues = new Dictionary<string, IEnumValue>
+            {
+                { "dll",             new PageEnumValue(new EnumValue {Name = "dll",    DisplayName = "0" }) },
+                { "exe",             new PageEnumValue(new EnumValue {Name = "exe",    DisplayName = "1" }) },
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: Help me understand how DisplayName works?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                { "appcontainerexe", new PageEnumValue(new EnumValue {Name = "exe",    DisplayName = "1" }) }
+            };
+
+            public bool AllowCustomValues => false;
+
+            public Task<ICollection<IEnumValue>> GetListedValuesAsync()
+            {
+                // CPS doesn't like it if we have duplicate values in the list of possible values. So take just
+                // the first three which are not duplicates.
+                return Task.FromResult<ICollection<IEnumValue>>(_outputTypeValues.Values.Take(3).ToList());
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: How often do these get called? This allocates like crazy. Any reason why ToList() and not ToArray()?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            }
+
+            public Task<IEnumValue> TryCreateEnumValueAsync(string userSuppliedValue)
+            {
+                if (_outputTypeValues.ContainsKey(userSuppliedValue))
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: Nit: TryGetValue to avoid the double-lookup.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 3 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Why we need to redefine properties that haven't changed?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 5 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Help me understand the displayname properties, these don't look like the displaynames but actually the underlying values of the output type.

**Comment by davkean** [comment]
:+1: The general approach looks sound, just would like to understand the DisplayName handling.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 5 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: Yes CP seems to use DisplayName as the underlying value of the enums. When the property pages query the browseobject for the value of the property they get back the DisplayName. IEnumValue doesn't have anything called Value in it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml (hunk 3 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/General.BrowseObject.xaml]: I just kept them all together so that if we eventually need to take ownership of this file we can simply just delete the copy in MSBuild.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs (hunk 10 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: Split _outputTypeValues into two so that I don't have to do a Take(3).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs (hunk 11 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/OutputTypeEnumProvider.cs]: Done

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj (hunk 8 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj]: Done. I remove the dot from the file so that I can have a class name for it.
